modules/map/tools/map_datachecker/server/channel_verify.cc

```diff
@@ -0,0 +1,225 @@
+    return return_state_;
+  }
+  incremental_check(records_path);
+  return_state_ = ErrorCode::SUCCESS;
+  return return_state_;
+}
+
+std::shared_ptr<std::vector<OneRecordChannelCheckResult>>
+ChannelVerify::get_check_result() const {
+  return sp_vec_check_result_;
+}
+
+int ChannelVerify::incremental_check(
+    const std::vector<std::string>& records_path) {
+
+  std::vector<std::string> not_check_records_path;
+  AINFO << "all records path:";
+  for (size_t i = 0; i < records_path.size(); ++i) {
+    AINFO << "[" << i << "]: " << records_path[i];
+    if (is_record_file(records_path[i]) &&
+        !is_record_checked(records_path[i])) {
+      not_check_records_path.push_back(records_path[i]);
+    }
+  }
+
+  AINFO << "not_check_records_path:";
+  for (size_t i = 0; i < not_check_records_path.size(); ++i) {
+    AINFO << "[" << i << "]: " << not_check_records_path[i];
+    OneRecordChannelCheckResult check_result =
+        check_record_channels(not_check_records_path[i]);
+    if (check_result.record_path.empty()) {
+      continue;
+    }
+    sp_vec_check_result_->push_back(check_result);
+  }
+
+  return 0;
+}
+
+bool ChannelVerify::is_record_file(const std::string& record_path) const {
+  if (!boost::filesystem::exists(record_path)) {
+    AINFO << "path [" << record_path << "] does not exist";
+    return false;
+  }
+  if (!boost::filesystem::is_regular_file(record_path)) {
+    AINFO << "path [" << record_path << "] is not a regular file";
+    return false;
+  }
+  // To avoid disk overhead caused by opening files twice, the real
+  // file checking is placed in the function [ChannelVerify::get_record_info]
+  return true;
+}
+
+std::vector<std::string> ChannelVerify::get_records_path(
+    const std::string& record_dir_or_record_full_path) const {
+  // record_dir_or_record_full_path is record fullpath or
+  // directory which contains some records
+  std::vector<std::string> records_path;
+  // 1. check record_dir_or_record_full_path is valid or not
+  boost::filesystem::path path(record_dir_or_record_full_path);
+  if (!boost::filesystem::exists(path)) {
+    AINFO << "record path [" << record_dir_or_record_full_path
+          << "] does not exist";
+    return records_path;
+  }
+
+  if (is_record_file(record_dir_or_record_full_path)) {
+    records_path.push_back(record_dir_or_record_full_path);
+  } else if (boost::filesystem::is_directory(path)) {
+    using dit_t = boost::filesystem::directory_iterator;
+    dit_t end;
+    for (dit_t it(record_dir_or_record_full_path); it != end; ++it) {
+      if (is_record_file(it->path().string())) {
+        records_path.push_back(it->path().string());
+      }
+    }
+  }
+  return records_path;
+}
+
+bool ChannelVerify::is_record_checked(const std::string& record_path) {
+  if (checked_records_.find(record_path) == checked_records_.end()) {
+    checked_records_.insert(record_path);
+    return false;
+  } else {
+    return true;
+  }
```

Comment:
Simply `return checked_records_.insert(record_path).second;`
https://en.cppreference.com/w/cpp/container/set/insert

Reply:
thank you so much